rust-runtime/aws-smithy-runtime-api/src/type_erasure.rs

```diff
@@ -62,75 +62,84 @@ where
     }
 }
 
-impl<T: 'static> Deref for TypedBox<T> {
+impl<T> fmt::Debug for TypedBox<T>
+where
+    T: fmt::Debug + Send + Sync + 'static,
+{
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        f.write_str("TypedBox:")?;
+        self.inner.downcast_ref::<T>().expect("typechecked").fmt(f)
+    }
+}
+
+impl<T: fmt::Debug + Send + Sync + 'static> Deref for TypedBox<T> {
     type Target = T;
 
     fn deref(&self) -> &Self::Target {
         self.inner.downcast_ref().expect("type checked")
     }
 }
 
-impl<T: 'static> DerefMut for TypedBox<T> {
+impl<T: fmt::Debug + Send + Sync + 'static> DerefMut for TypedBox<T> {
     fn deref_mut(&mut self) -> &mut Self::Target {
         self.inner.downcast_mut().expect("type checked")
     }
 }
 
-#[derive(Debug)]
-pub struct TypedRef<'a, T> {
-    inner: &'a TypeErasedBox,
-    _phantom: PhantomData<T>,
-}
-
-impl<'a, T: 'static> TypedRef<'a, T> {
-    pub fn assume_from(type_erased: &'a TypeErasedBox) -> Option<TypedRef<'a, T>> {
-        if type_erased.downcast_ref::<T>().is_some() {
-            Some(TypedRef {
-                inner: type_erased,
-                _phantom: Default::default(),
-            })
-        } else {
-            None
-        }
-    }
+/// A new-type around `Box<dyn Debug + Send + Sync>`
+pub struct TypeErasedBox {
+    field: Box<dyn Any + Send + Sync>,
+    #[allow(dead_code)]
+    type_name: &'static str,
+    #[allow(clippy::type_complexity)]
+    debug: Box<dyn Fn(&TypeErasedBox, &mut fmt::Formatter<'_>) -> fmt::Result + Send + Sync>,
 }
 
-impl<'a, T: 'static> Deref for TypedRef<'a, T> {
-    type Target = T;
-
-    fn deref(&self) -> &Self::Target {
-        self.inner.downcast_ref().expect("type checked")
+impl fmt::Debug for TypeErasedBox {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        (self.debug)(self, f)
     }
 }
 
-/// A new-type around `Box<dyn Any + Send + Sync>`
-#[derive(Debug)]
-pub struct TypeErasedBox {
-    inner: Box<dyn Any + Send + Sync>,
-}
-
 impl TypeErasedBox {
-    // Creates a new `TypeErasedBox`.
-    pub fn new(inner: Box<dyn Any + Send + Sync>) -> Self {
-        Self { inner }
+    pub fn new<T: Send + Sync + fmt::Debug + 'static>(value: T) -> Self {
+        let debug = |value: &TypeErasedBox, f: &mut fmt::Formatter<'_>| {
+            f.write_str("TypeErasedBox:")?;
```

Comment:
I would probably remove this from here—if you want you can re-add it directly in the debug impl. Then you can reuse this function in typedbox